### PR TITLE
Generate Cython code at build time

### PR DESCRIPTION
**Pull request by yunline**
- [x] Generate Cython code at build time.
- [x] Remove `cython` option from setup.py. (but keep the `cython_only`)
- [x] Add cython-generated files into .gitignore .
- [x] #2395

Reference: 
[https://github.com/pygame/pygame/pull/2916](https://github.com/pygame/pygame/pull/2916)
#1862

### CODE REVIEW

File: .github/workflows/build-windows.yml [hunk 1]
```diff
@@ -156,7 +158,7 @@ jobs:
           set DISTUTILS_USE_SDK=1
           set MSSdk=1
           python -m pip install setuptools wheel requests numpy Sphinx
-          python setup.py docs
+          python3 setup.py docs
```
Comment (Starbuck5) [inline .github/workflows/build-windows.yml:161]: What, why?

File: .github/workflows/build-windows.yml [hunk 2]
```diff
@@ -131,6 +131,8 @@ jobs:
       CIBW_BUILD: ${{ matrix.pyversions }}
       CIBW_ARCHS: ${{ matrix.winarch }}
 
+      CIBW_BEFORE_BUILD: python -m pip install setuptools wheel Cython requests
+
```
Comment (Starbuck5) [inline .github/workflows/build-windows.yml:135]: Why not just move the cython install into the line 158/160 pip install?

File: .github/workflows/build-windows.yml (hunk 1 above)
Comment (yunline) [reply .github/workflows/build-windows.yml:161]: ???
I will change it back

File: .github/workflows/build-windows.yml (hunk 2 above)
Comment (yunline) [reply .github/workflows/build-windows.yml:135]: If moved, the CIBW would not run correctly ( LNK1104 error )

Comment (Starbuck5) [reply .github/workflows/build-windows.yml:135]: That's crazy.

**Review by MyreMylar — APPROVED** [review]
I'm generally approving of this as a stop gap before steadily turning cython files to C files.

**Comment by pmp-p** [comment]
Though the wasm CI is already doing that, this is probably a bad idea for testing against python git where cython files **may** need to be edited.

This is what i do locally after 3.11 and 3.12(git) are built:

- i build with python3.11 : cython files + pygame
- Then if i know cython files will fail on 3.12, i patch the 3.11 generated one
- then continue build without running cython_only.

**Comment by yunline** [comment]
> Though the wasm CI is already doing that, this is probably a bad idea for testing against python git where cython files **may** need to be edited.

Maybe we can solve this problem by adding a 'non-cython' option.

**Comment by MyreMylar** [comment]
slightly longer term (but getting closer) the plan is to clear out all the cython code entirely.

Is this PR useful enough to current cython development to get over the downsides? Or should we close it. I tend to avoid cython entirely.

File: setup.py [hunk 3]
```diff
@@ -288,11 +279,12 @@ def consume_arg(name):
...
         cythonize_one(**kwargs)
-    
-    if cython_only:
-        sys.exit(0)
 
-no_compilation = any(x in ['lint', 'format', 'docs'] for x in sys.argv)
+no_compilation = any(x in ['lint', 'format', 'docs','cython_only'] for x in sys.argv)
+
+if not no_compilation: # Compile cython before compilation
+    compile_cython()
```
Comment (dr0id) [inline setup.py:286]: Why are those lines not in the build command? Or are these files needed elsewhere too?

File: setup.py [hunk 4]
```diff
@@ -947,6 +939,16 @@ class LintCommand(LintFormatCommand):
 class FormatCommand(LintFormatCommand):
     format = True
 
+@add_command("cython_only")
+class CythonOnlyCommand(Command):
+
+    user_options = []
+    def initialize_options(self):pass
```
Comment (dr0id) [inline setup.py:946]: please add a description:
`    description = "short doc what this command does"`

can be seen with: `python setup.py --help-commands`

there is also `python setup.py cython_only --help`
not sure what that would produce at the moment.

Comment (yunline) [reply setup.py:946]: image
It seems other commands don't have descriptions too. Maybe we need another pr to fix that.

Comment (dr0id) [reply setup.py:929]: yes, see https://github.com/pygame-community/pygame-ce/pull/2360

**Comment by ankith26** [comment]
This PR has been superseded by #2831 because we recently introduced the new buildconfig, so I'm closing this 

But thanks for starting the work on this!
